### PR TITLE
Using new ppm function blocks

### PR DESCRIPTION
**Pull request by KaushikMalapati**
## Description

Using the new PPM function blocks from https://github.com/pcdshub/lcls-twincat-common-components/pull/69 and updating IM4K2's custom ppm function blocks to similarly calculate ppm voltage using responsivities

## Motivation and Context

https://jira.slac.stanford.edu/browse/ECS-875

## How Has This Been Tested?

This will be tested around October 19-20th

## Where Has This Been Documented?

N/A

## Pre-merge checklist
- [ ] Code works interactively
- [ ] Code contains descriptive comments
- [ ] Test suite passes locally
- [ ] Libraries are set to fixed versions and not ``Always Newest``
- [ ] Code committed with ``pre-commit`` (alternatively ``pre-commit run --all-files``)

### CODE REVIEW

File: plc-kfe-rix-motion/kfe_rix_motion/POUs/IM4K2/FB_PPM_PowerMeter_IM4K2.TcPOU [hunk 1]
```diff
@@ -20,87 +41,289 @@ VAR
...
+// Getting frequency based on accelerator source
+(*
+CASE uAccelerator OF
+    0: uFrequency := uNCFrequency;
+    1: uFrequency := uSCFrequency;
+    2: uFrequency := 0; // ASTA?
+ELSE
+    uFrequency := 0;
+END_CASE*)
+
```
Comment (KaushikMalapati) [inline plc-kfe-rix-motion/kfe_rix_motion/POUs/IM4K2/FB_PPM_PowerMeter_IM4K2.TcPOU:280]: You can see how I am using defines to determine xray line here. Presumably after testing with beam on rix motion, I will open another pr for common components that has this change and any others I need.

**Comment by KaushikMalapati** [comment]
This is the using common-componenets post https://github.com/pcdshub/lcls-twincat-common-components/pull/69, with a change - instead of doing `PMPS_GVL.g_areVBoundaries = PMPS_GVL.g_areVBoundariesL;` to check the xray line, I am using `IF defined (K/L}`. This is because when I was checking objects I got errors for trying to check for array equality. I think I added those lines after I tested on rix-motion the first time and assumed array equality would work instead of using the defineds. I also set the version to 3.8.0 for now since that would be the next version of cc.

Other changes are 

- adding responsivities for all ppm fbs
- regenerating iocboot
- copying the new PPM and PPM_PowerMeter function blocks into IM4K2 but keeping the modifications for its high precision adc, which I think are

- adding back all the PPM_PowerMeter parameters when calling the function block
- using DINT for iVoltage
- using FB_ThermoCouple instead of TempSensor (although it looks unused anyways)
- using the custom FB_AnalogInput_IM4K2.

**Review by ZLLentz — COMMENTED** [review]
I think all the code looks good and the responsivity settings match the ticket 1:1

File: plc-kfe-rix-motion/kfe_rix_motion/POUs/IM4K2/FB_PPM_PowerMeter_IM4K2.TcPOU [hunk 2]
```diff
@@ -20,87 +41,289 @@ VAR
...
+    ELSIF uSXR_Accelerator = 1 THEN
+        fFrequency := uSXR_SC_Frequency;
+    END_IF
+{ELSIF defined (L)}
+    IF uHXR_Accelerator = 0 THEN
+        fFrequency := fHXR_NC_Frequency;
+    ELSIF uHXR_Accelerator = 1 THEN
+        fFrequency := uHXR_SC_Frequency;
+    END_IF
+{END_IF}
```
Comment (KaushikMalapati) [inline plc-kfe-rix-motion/kfe_rix_motion/POUs/IM4K2/FB_PPM_PowerMeter_IM4K2.TcPOU:293]: This is how I use the defineds here and in my locally built common components. I made sure that it builds without error.

**Comment by ZLLentz** [comment]
> instead of doing PMPS_GVL.g_areVBoundaries = PMPS_GVL.g_areVBoundariesL; to check the xray line, I am using IF defined (K/L}

Maybe this becomes another small PR and then I tag the library?

File: plc-kfe-rix-motion/kfe_rix_motion/kfe_rix_motion.plcproj [hunk 3]
```diff
@@ -191,7 +191,7 @@
       <Resolution>LCLS General, 2.10.0 (SLAC)</Resolution>
     </PlaceholderResolution>
     <PlaceholderResolution Include="lcls-twincat-common-components">
-      <Resolution>lcls-twincat-common-components, 3.4.0 (SLAC)</Resolution>
+      <Resolution>lcls-twincat-common-components, 3.8.0 (SLAC)</Resolution>
```
[the remark this answers is not in the document]
Comment (KaushikMalapati) [reply plc-kfe-rix-motion/kfe_rix_motion/kfe_rix_motion.plcproj:194]: Not sure what you mean by deploy on the prog node. I changed the version of my local common-components repository, saved and installed as a library, then installed on kfe-rix-motion and changed the placeholder version to match. Is that the same thing or is there a different way to update libraries?

**Comment by KaushikMalapati** [comment]
> > instead of doing PMPS_GVL.g_areVBoundaries = PMPS_GVL.g_areVBoundariesL; to check the xray line, I am using IF defined (K/L}
> 
> Maybe this becomes another small PR and then I tag the library?

Do I keep the version at 3.8.0 and set it to released for that or should I make it 0.0.0?

**Comment by ZLLentz** [comment]
Keep it at 3.8.0 and we'll make the tag 3.8.0, then we can test this again after the tag

**Comment by ZLLentz** [comment]
3.8.0 is now installed on the programming node
Sometimes you need to close/re-open twincat to get new libraries to appear

**Comment by KaushikMalapati** [comment]
Since I installed my local repo as 3.8.0 before you released it on prog, what do I do to make sure I am using the actual release? Also, did you something like the steps described in https://confluence.slac.stanford.edu/display/PCDS/How+to+Release+and+Install+Vacuum+PLC+Library to make the release?

**Comment by ZLLentz** [comment]
> Since I installed my local repo as 3.8.0 before you released it on prog

I'm not sure what you mean by this, but

> what do I do to make sure I am using the actual release?

I would use the TwinCAT library browser and verify that the files are what you expect. If what you were running off of previously matches what's there, we're probably OK regardless of which version is "officially" there.

In the future, please never install a library that hasn't been released. This can cause lots of issues if we're being too ambitious with our updates (in this case, probably harmless).

> Also, did you something like the steps described in https://confluence.slac.stanford.edu/display/PCDS/How+to+Release+and+Install+Vacuum+PLC+Library to make the release?

Sort of, I don't like downloading and transferring the zip. I ran tc-release like in the document there but then I used git to get the v3.8.0 source code onto the node, then save and install.

**Comment by KaushikMalapati** [comment]
> In the future, please never install a library that hasn't been released. This can cause lots of issues if we're being too ambitious with our updates (in this case, probably harmless).

I will keep that in mind, but if you are working on a library update, how would you try it before releasing it without installing it?

**Comment by ZLLentz** [comment]
> if you are working on a library update, how would you try it before releasing it without installing it?

It should usually be ok to install at `0.0.0`, and then you won't break someone who is running "latest", you will only clobber someone else trying to do the same thing.

You can also rename the library and test with the renamed version (lcls-twincat-motion-zlentz, for example)

**Review by ZLLentz — APPROVED** [review]
whoops, forgot to drop my approval

**Comment by ZLLentz** [comment]
btw, I'm always open to ideas for improving these release workflows, I realize the way we have it now might not be ideal for various reasons

**Review by ZLLentz — APPROVED** [review]

**Pull request merged**